### ISSUE TITLE
Running Tests with memory checker has defects

### ISSUE THREAD

**Issue by TomzBench** [benmcollins/libjwt#114]
It looks like the test suite is having memory leaks...  I run the check test suite using CTest (from CMake).  CTest has an option to run these tests while checking for memory errors by running the following command ```ctest -T memcheck```  ...It looks like the tests are leaking.

Even though these are only tests, it's best to fix memory errors in the test because the memory errors reported could be potentially in the production code and not in the test code. But it's not possible for the test report to distinguish between the two. So it's best to make sure the tests do not have memory errors.

Here is the output:

```
   Site: localhost.localdomain
   Build name: Linux-c++

    Start 1: jwt_dump
1/8 MemCheck #1: jwt_dump .........................   Passed    0.65 sec
    Start 2: jwt_ec
2/8 MemCheck #2: jwt_ec ...........................   Passed    3.98 sec
    Start 3: jwt_encode
3/8 MemCheck #3: jwt_encode .......................   Passed    1.06 sec
    Start 4: jwt_grant
4/8 MemCheck #4: jwt_grant ........................   Passed    0.75 sec
    Start 5: jwt_header
5/8 MemCheck #5: jwt_header .......................   Passed    0.70 sec
    Start 6: jwt_new
6/8 MemCheck #6: jwt_new ..........................   Passed    1.64 sec
    Start 7: jwt_rsa
7/8 MemCheck #7: jwt_rsa ..........................   Passed    6.34 sec
    Start 8: jwt_validate
8/8 MemCheck #8: jwt_validate .....................   Passed    1.20 sec

100% tests passed, 0 tests failed out of 8

Total Test time (real) =  16.34 sec
-- Processing memory checking output:
1/8 MemCheck: #1: jwt_dump .........................   Defects: 187
2/8 MemCheck: #2: jwt_ec ...........................   Defects: 537
3/8 MemCheck: #3: jwt_encode .......................   Defects: 325
4/8 MemCheck: #4: jwt_grant ........................   Defects: 277
5/8 MemCheck: #5: jwt_header .......................   Defects: 277
6/8 MemCheck: #6: jwt_new ..........................   Defects: 717
7/8 MemCheck: #7: jwt_rsa ..........................   Defects: 655
8/8 MemCheck: #8: jwt_validate .....................   Defects: 537
MemCheck log files can be found here: ( * corresponds to test number)

Memory checking results:
Potential Memory Leak - 3512
```

**Comment by benmcollins** [comment]
I'm very skeptical of this. I use the auto tools to build and it already runs the tests through valgrind in several ways and produces no leaks of any kind. I've never released LibJWT with known test failures, including memory leaks.

Maybe you should locate these supposed memory leaks because I don't use CMake, and I'm not seeing them with valgrind.

**Comment by TomzBench** [comment]
I checked again and it appears on my machine to be reporting memory defects as "still reachable".  To be sure I also build with autotools and ran the test binary with valgrind outside of the test suite using the command:

```
valgrind --tool=memcheck --leak-check=full -s ./tests/jwt_dump
```

Both ```make check``` (autotools), ```ctest -T memcheck``` (cmake), and running the binary with valgrind directly using the command above, report unfreed memory as "still reachable".  Valgrind notes this about "Still reachable"

> "still reachable" means your program is probably ok -- it didn't free some memory it could have. This is quite common and often reasonable. Don't use --show-reachable=yes if you don't want to see these reports. "suppressed" means that a leak error has been suppressed.

So it doesn't seem like a big deal - however the CMake test suite presentation layer dramatizes the items by labelling them as defects.  This then makes finding more significant memory errors hard to parse should some other error exist during development.  So I think the solution would be to some how make CMake not label "still reachable" as a defect - or - free the "still reachable" memory.

**Comment by benmcollins** [comment]
Sounds like something in the tests isn't calling jwt_free(). I'll certainly recheck this.

Thanks.

**Comment by benmcollins** [comment]
I believe this to be fixed in v1.14.0
